pkg/providers/vsphere/vmprovider_vm.go

```diff
@@ -1434,41 +1422,113 @@ func (vs *vSphereVMProvider) getVMYamlFromSnapshot(
-		vmCtx.Logger.Info("Skipping snapshot creation because the VirtualMachineSnapshot is being deleted")
+	vmCtx.Logger.Info("Found snapshots for VM", "snapshotCount", len(vmSnapshots), "snapshots", vmSnapshots)
+
+	// If no snapshots found, nothing to do
+	if len(vmSnapshots) == 0 {
+		return nil
+	}
+
+	// Sort snapshots by creation timestamp to ensure consistent ordering
+	sort.Slice(vmSnapshots, func(i, j int) bool {
+		return vmSnapshots[i].CreationTimestamp.Before(&vmSnapshots[j].CreationTimestamp)
+	})
+
+	// Find the first (oldest) snapshot that is not ready
+	// We process snapshots in chronological order, one at a time
+	// Also calculate the total pending snapshots so we can requeue the
+	// request to continue processing other snapshots.
+	var snapshotToProcess *vmopv1.VirtualMachineSnapshot
+	totalPendingSnapshots := 0
+	for _, snapshot := range vmSnapshots {
+		// Skip snapshots that are already ready
+		if pkgcnd.IsTrue(&snapshot, vmopv1.VirtualMachineSnapshotReadyCondition) {
+			vmCtx.Logger.V(4).Info("Skipping snapshot that is already ready",
+				"snapshotName", snapshot.Name)
+			continue
+		}
+
+		if snapshot.DeletionTimestamp.IsZero() {
+			totalPendingSnapshots++
+		}
+
+		// Found the first snapshot that needs processing
+		if snapshotToProcess == nil {
+			snapshotToProcess = &snapshot
+		}
+	}
+
+	// If no snapshot needs processing, we're done
+	if snapshotToProcess == nil {
+		vmCtx.Logger.V(4).Info("All snapshots are ready or being deleted, nothing to process")
+		return nil
+	}
+
+	// If the oldest snapshot is being deleted, wait for it to finish.
+	if !snapshotToProcess.DeletionTimestamp.IsZero() {
+		vmCtx.Logger.V(4).Info("Snapshot is being deleted, waiting for completion",
+			"snapshotName", snapshotToProcess.Name)
 		return nil
 	}
 
+	// If the oldest snapshot is in progress, wait for it to complete
+	if pkgcnd.IsTrue(snapshotToProcess, "VirtualMachineSnapshotInProgress") {
+		vmCtx.Logger.V(4).Info("Snapshot is in progress, waiting for completion",
+			"snapshotName", snapshotToProcess.Name)
+		return nil
+	}
+
+	// Mark the snapshot as in progress before starting the operation to
+	// prevent concurrent snapshots.
+	if err := vs.markSnapshotInProgress(vmCtx, snapshotToProcess); err != nil {
+		vmCtx.Logger.Error(err, "Failed to mark snapshot as in progress", "snapshotName", snapshotToProcess.Name)
+		return err
+	}
+
 	snapArgs := virtualmachine.SnapshotArgs{
 		VMCtx:      vmCtx,
 		VcVM:       vcVM,
-		VMSnapshot: vmSnapshot,
+		VMSnapshot: *snapshotToProcess,
 	}
 
-	vmCtx.Logger.Info("Creating a new snapshot of the virtual machine")
+	vmCtx.Logger.Info("Creating snapshot on vSphere",
+		"snapshotName", snapshotToProcess.Name)
 	snapMoRef, err := virtualmachine.SnapshotVirtualMachine(snapArgs)
 	if err != nil {
-		vmCtx.Logger.Error(err, "failed to snapshot virtual machine")
+		// Mark the snapshot as failed and clear in-progress status
+		vs.markSnapshotFailed(vmCtx, snapshotToProcess, err)
```

Comment:
If `marking failed` Patch call failed, there is a chance that the snapshot stuck in `inprogress` condition forever. Maybe we should add retry around the patch call?

Reply:
But that would just be a temporary workaround. If the APIserver is down for a long time, we'd better to have a timeout check on the in progress status, in case the retry doesn't cover that amount of time.
But that could be a future improvement